File: ios/SecurityCheckerModule.swift
import ExpoModulesCore

public class SecurityCheckerModule: Module, VulnerabilitiesDetectorDelegate {
    
    let vulnerabilitiesDetector = VulnerabilitiesDetector()
    
    
    func onChange(report: VulnerabilitiesReport) {
        do {
            let encoded = try report.toDictionary()
            self.sendEvent("onChange", encoded)
        } catch {
            print("Error encoding report: \(error)")
        }
    }
    
    
    public func definition() -> ModuleDefinition {
        Name("SecurityChecker")
        
        OnCreate {
            self.vulnerabilitiesDetector.delegate = self
        }
        
        Events("onChange")
        
        Function("check") {
            let report = vulnerabilitiesDetector.runDetection()

            do {
                return try report.toDictionary()
            } catch {
                throw Exception(name: "EncodingError", description: error.localizedDescription)
            }
        }
    }
}

File: android/src/main/java/expo/modules/securitychecker/SecurityCheckerModule.kt
package expo.modules.securitychecker

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers

class SecurityCheckerModule : Module(), VulnerabilitiesDetectorDelegate {
  private val vulnerabilitiesDetector = VulnerabilitiesDetector()
  private val coroutineScope = CoroutineScope(Dispatchers.Default)

  override fun definition() = ModuleDefinition {
    Name("SecurityChecker")

    Events("onChange")

    OnCreate {
      vulnerabilitiesDetector.delegate = this@SecurityCheckerModule
    }

    Function("check") {
      val report = vulnerabilitiesDetector.runDetection()
      report.toJSMap()
    }
  }

  override fun onChange(report: VulnerabilitiesReport) {
    try {
      val encoded = report.toJSMap()
      sendEvent("onChange", encoded)
    } catch (e: Exception) {
      println("Error serializing report: ${e.message}")
    }
  }
}
